### ISSUE TITLE
VS2019 suggests null-coalescing operator rejected by compiler

### ISSUE THREAD

**Issue by Zannick** [dotnet/roslyn#47317]
**Version Used**: 16.7.2

<details>
<summary>Full Versions</summary>

```
Microsoft Visual Studio Community 2019 Version 16.7.2
VisualStudio.16.Release/16.7.2+30413.136
Microsoft .NET Framework Version 4.8.04084
Installed Version: Community

ASP.NET and Web Tools 2019   16.7.532.28833
ASP.NET Core Razor Language Services   16.1.0.2035807+72d099b977d3a85e65fa3b0614ca8cfc803fef02
ASP.NET Web Frameworks and Tools 2019   16.7.532.28833
Azure App Service Tools v3.0.0   16.7.532.28833
Azure Functions and Web Jobs Tools   16.7.532.28833
C# Tools   3.7.0-6.20412.3+d3c3a44a4e7ad31cc75c59be0d3df4a19ff33878
Common Azure Tools   1.10
```
</details>

**Steps to Reproduce**:

Case 1:
Code includes nullable object and the `??` operator used to instantiate the object once:
```
private List<Type> enums;
public List<Type> Types
{
    get { return enums ?? (enums = new List<Type>()); }
}
```
VS suggests IDE0074 to change getter to `{ return enums ??= new List<Type>(); }`

Case 2:
Write invalid code using the `??` operator on non-nullable types.
```
private int a;
private int b;
... eg. in constructor: ...
    a = b ?? (b = 1);
```
VS reports `CS0019: Operator '??' cannot be applied to operands of type 'int' and 'int'` 

**Expected Behavior**:

Case 1. The suggestion is not given **OR** the code compiles after accepting the suggestion.
Case 2. The suggestion is not offered on this code.

**Actual Behavior**:

Case 1. The suggestion is offered and the code fails with `error CS8652: The feature 'coalescing assignment' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.`
Case 2. The incorrect code has the IDE0074 suggestion to make it `a = b ??= 1`. Applying the suggestion causes compilation to fail with error CS8652, while the line is still annotated with `CS0019: Operator '??=' cannot be applied to operands of type 'int' and 'int'`

**Comment by CyrusNajmabadi** [comment]
> Case 2. The suggestion is not offered on this code.

I don't see why not.  The code is invalid before.  It remains equally invalid after.  In general we don't limit our features in invalid code, since invalid cod is the *norm* during editing sessions.  For all we know, you're going to change it to an `int?` later.  But we don't want to force you to have to make that change first before offering the simplification that would valid for `x ?? (x = ...)`

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  @333fred do you know why trying to use `??=` results in:

```
CS8652: The feature 'coalescing assignment' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
```

I would have thought this feature was well out of preview.

**Comment by CyrusNajmabadi** [comment]
Oh, nevermind.  User is on an old version of VS.  This was fixed with https://github.com/dotnet/roslyn/pull/44798.  Closing out :)

**Comment by CyrusNajmabadi** [comment]
> while the line is still annotated with CS0019

Just to be clear on this point.  Code transformations are not expected or intended to just fix errors.  They are often to help people manipulate code *regardless* of the current error state.  This falls into that category.  If you have `x ?? (x = y)` we offer to covert that to `x ??= y`.  We do that when the current code is legal and the transformation does not change semantics.  And we offer it when the current code is illegal as you may be in the middle of work and want this change prior to fixing up other things.

**Comment by Zannick** [comment]
> I don't see why not. The code is invalid before. It remains equally invalid after. In general we don't limit our features in invalid code, since invalid cod is the _norm_ during editing sessions.

Good point, thanks. :)

> User is on an old version of VS.

Haha, and I even updated to the latest first! :p

Thanks!

**Comment by RikkiGibson** [comment]
The fix #44798 should be in latest VS 16.7, going by:

```sh
> git merge-base --is-ancestor 6b3ce16 upstream/release/dev16.7; echo $?
0
```

i.e. our VS 16.7 release branch contains the commit 6b3ce16
